test/strings/basic.jl

```diff
@@ -428,6 +428,9 @@ foobaz(ch) = reinterpret(Char, typemax(UInt32))
 # issue #17624, missing getindex method for String
 @test "abc"[:] == "abc"
 
-# PR #18259 reverted because failing test below
-a = [x for x in String([0xcf, 0x83, 0x83, 0x83, 0x83])]
-@test a[1] == 'σ'
+# we assume for performance that next/nextind return values past the end of a
+# string's underlying data; this helps with performance of `done`.
```

Comment:
Throwing an exception on iterating over characters in an invalid string seems like an okay behavior to me.

Basically, if you are in a situation where you are stuffing arbitrary binary data into a `String` for some reason, then you usually shouldn't be iterating over characters at all.

(One thing that I find annoying is that `show` barfs with an error on showing an arbitrary string, whereas I think it should give some useful information, at least in the REPL.  But that is something for a different PR.)